File: go/src/github.com/flanglet/kanzi/test/TestEntropyCodec.go
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	kanzi "github.com/flanglet/kanzi"
	"github.com/flanglet/kanzi/bitstream"
	"github.com/flanglet/kanzi/entropy"
	"github.com/flanglet/kanzi/io"
	"math/rand"
	"os"
	"strings"
	"time"
)

func main() {
	var name = flag.String("type", "ALL", "Type of codec (all, Huffman, ANS, Range, FPAQ, CM, PAQ, TPAQ, ExpGolomg or RiceGolomb)")

	// Parse
	flag.Parse()
	name_ := strings.ToUpper(*name)

	if name_ == "ALL" {
		fmt.Printf("\n\nTestHuffmanCodec")
		TestCorrectness("HUFFMAN")
		TestSpeed("HUFFMAN")
		fmt.Printf("\n\nTestANS0Codec")
		TestCorrectness("ANS0")
		TestSpeed("ANS0")
		fmt.Printf("\n\nTestANS1Codec")
		TestCorrectness("ANS1")
		TestSpeed("ANS1")
		fmt.Printf("\n\nTestRangeCodec")
		TestCorrectness("RANGE")
		TestSpeed("RANGE")
		fmt.Printf("\n\nTestFPAQEntropyCoder")
		TestCorrectness("FPAQ")
		TestSpeed("FPAQ")
		fmt.Printf("\n\nTestCMEntropyCoder")
		TestCorrectness("CM")
		TestSpeed("CM")
		fmt.Printf("\n\nTestPAQEntropyCoder")
		TestCorrectness("PAQ")
		TestSpeed("PAQ")
		fmt.Printf("\n\nTestTPAQEntropyCoder")
		TestCorrectness("TPAQ")
		TestSpeed("TPAQ")
		fmt.Printf("\n\nTestExpGolombCodec")
		TestCorrectness("EXPGOLOMB")
		TestSpeed("EXPGOLOMB")
		fmt.Printf("\n\nTestRiceGolombCodec")
		TestCorrectness("RICEGOLOMB")
		TestSpeed("RICEGOLOMB")
	} else if name_ != "" {
		fmt.Printf("\n\nTest%vCodec", name_)
		TestCorrectness(name_)
		TestSpeed(name_)
	}
}

func getPredictor(name string) kanzi.Predictor {
	switch name {
	case "PAQ":
		res, _ := entropy.NewPAQPredictor()
		return res

	case "FPAQ":
		res, _ := entropy.NewFPAQPredictor()
		return res

	case "TPAQ":
		res, _ := entropy.NewTPAQPredictor(nil)
		return res

	case "CM":
		res, _ := entropy.NewCMPredictor()
		return res

	default:
		panic(fmt.Errorf("Unsupported type: '%s'", name))
	}
}

func getEncoder(name string, obs kanzi.OutputBitStream) kanzi.EntropyEncoder {
	switch name {
	case "PAQ":
		res, _ := entropy.NewBinaryEntropyEncoder(obs, getPredictor(name))
		return res

	case "FPAQ":
		res, _ := entropy.NewBinaryEntropyEncoder(obs, getPredictor(name))
		return res

	case "TPAQ":
		res, _ := entropy.NewBinaryEntropyEncoder(obs, getPredictor(name))
		return res

	case "CM":
		res, _ := entropy.NewBinaryEntropyEncoder(obs, getPredictor(name))
		return res

	case "HUFFMAN":
		res, _ := entropy.NewHuffmanEncoder(obs)
		return res

	case "ANS0":
		res, _ := entropy.NewANSRangeEncoder(obs, 0)
		return res

	case "ANS1":
		res, _ := entropy.NewANSRangeEncoder(obs, 1)
		return res

	case "RANGE":
		res, _ := entropy.NewRangeEncoder(obs)
		return res

	case "EXPGOLOMB":
		res, _ := entropy.NewExpGolombEncoder(obs, true)
		return res

	case "RICEGOLOMB":
		res, _ := entropy.NewRiceGolombEncoder(obs, true, 4)
		return res

	default:
		panic(fmt.Errorf("No such entropy encoder: '%s'", name))
	}

	return nil
}

func getDecoder(name string, ibs kanzi.InputBitStream) kanzi.EntropyDecoder {
	switch name {
	case "PAQ":
		pred := getPredictor(name)

		if pred == nil {
			panic(fmt.Errorf("No such entropy decoder: '%s'", name))
		}

		res, _ := entropy.NewBinaryEntropyDecoder(ibs, pred)
		return res

	case "FPAQ":
		pred := getPredictor(name)

		if pred == nil {
			panic(fmt.Errorf("No such entropy decoder: '%s'", name))
		}

		res, _ := entropy.NewBinaryEntropyDecoder(ibs, pred)
		return res

	case "TPAQ":
		pred := getPredictor(name)

		if pred == nil {
			panic(fmt.Errorf("No such entropy decoder: '%s'", name))
		}

		res, _ := entropy.NewBinaryEntropyDecoder(ibs, pred)
		return res

	case "CM":
		pred := getPredictor(name)

		if pred == nil {
			panic(fmt.Errorf("No such entropy decoder: '%s'", name))
		}

		res, _ := entropy.NewBinaryEntropyDecoder(ibs, pred)
		return res

	case "HUFFMAN":
		res, _ := entropy.NewHuffmanDecoder(ibs)
		return res

	case "ANS0":
		res, _ := entropy.NewANSRangeDecoder(ibs, 0)
		return res

	case "ANS1":
		res, _ := entropy.NewANSRangeDecoder(ibs, 1)
		return res

	case "RANGE":
		res, _ := entropy.NewRangeDecoder(ibs)
		return res

	case "EXPGOLOMB":
		res, _ := entropy.NewExpGolombDecoder(ibs, true)
		return res

	case "RICEGOLOMB":
		res, _ := entropy.NewRiceGolombDecoder(ibs, true, 4)
		return res

	default:
		panic(fmt.Errorf("No such entropy decoder: '%s'", name))
	}

	return nil
}

func TestCorrectness(name string) {
	fmt.Printf("\n\nCorrectness test for %v\n", name)

	// Test behavior
	for ii := 1; ii < 20; ii++ {
		fmt.Printf("\n\nTest %v", ii)
		var values []byte
		rand.Seed(time.Now().UTC().UnixNano())

		if ii == 3 {
			values = []byte{0, 0, 32, 15, -4 & 0xFF, 16, 0, 16, 0, 7, -1 & 0xFF, -4 & 0xFF, -32 & 0xFF, 0, 31, -1 & 0xFF}
		} else if ii == 2 {
			values = []byte{0x3d, 0x4d, 0x54, 0x47, 0x5a, 0x36, 0x39, 0x26, 0x72, 0x6f, 0x6c, 0x65, 0x3d, 0x70, 0x72, 0x65}
		} else if ii == 4 {
			values = []byte{65, 71, 74, 66, 76, 65, 69, 77, 74, 79, 68, 75, 73, 72, 77, 68, 78, 65, 79, 79, 78, 66, 77, 71, 64, 70, 74, 77, 64, 67, 71, 64}
		} else if ii == 1 {
			values = make([]byte, 32)

			for i := range values {
				values[i] = byte(2) // all identical
			}
		} else if ii == 5 {
			values = make([]byte, 32)

			for i := range values {
				values[i] = byte(2 + (i & 1)) // 2 symbols
			}
		} else {
			values = make([]byte, 32)

			for i := range values {
				values[i] = byte(64 + 3*ii + rand.Intn(ii+1))
			}
		}

		fmt.Printf("\nOriginal: \n")

		for i := range values {
			fmt.Printf("%d ", values[i])
		}

		println()
		fmt.Printf("\nEncoded: \n")
		var bs io.BufferStream
		obs, _ := bitstream.NewDefaultOutputBitStream(&bs, 16384)
		dbgbs, _ := bitstream.NewDebugOutputBitStream(obs, os.Stdout)
		dbgbs.ShowByte(true)
		//dbgbs.Mark(true)
		ec := getEncoder(name, dbgbs)

		if ec == nil {
			os.Exit(1)
		}

		if _, err := ec.Encode(values); err != nil {
			fmt.Printf("Error during encoding: %s", err)
			os.Exit(1)
		}

		ec.Dispose()
		dbgbs.Close()
		println()
		fmt.Printf("\nDecoded: \n")

		ibs, _ := bitstream.NewDefaultInputBitStream(&bs, 16384)
		ed := getDecoder(name, ibs)

		if ed == nil {
			os.Exit(1)
		}

		ok := true
		values2 := make([]byte, len(values))

		if _, err := ed.Decode(values2); err != nil {
			fmt.Printf("Error during decoding: %s", err)
			os.Exit(1)
		}

		ed.Dispose()

		for i := range values2 {
			fmt.Printf("%v ", values2[i])

			if values[i] != values2[i] {
				ok = false
			}
		}

		if ok == true {
			fmt.Printf("\nIdentical")
		} else {
			fmt.Printf("\n! *** Different *** !")
			os.Exit(1)
		}

		ibs.Close()
		bs.Close()
		println()
	}
}

func TestSpeed(name string) {
	fmt.Printf("\n\nSpeed test for %v\n", name)
	repeats := []int{3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 9, 3}

	for jj := 0; jj < 3; jj++ {
		fmt.Printf("\nTest %v\n", jj+1)
		delta1 := int64(0)
		delta2 := int64(0)
		iter := 100
		size := 500000
		values1 := make([]byte, size)
		values2 := make([]byte, size)
		var bs io.BufferStream

		for ii := 0; ii < iter; ii++ {
			idx := jj

			for i := 0; i < size; i++ {
				i0 := i

				length := repeats[idx]
				idx = (idx + 1) & 0x0F
				b := byte(rand.Intn(256))

				if i0+length >= size {
					length = size - i0 - 1
				}

				for j := i0; j < i0+length; j++ {
					values1[j] = b
					i++
				}
			}

			obs, _ := bitstream.NewDefaultOutputBitStream(&bs, uint(size))
			ec := getEncoder(name, obs)

			if ec == nil {
				os.Exit(1)
			}

			// Encode
			before1 := time.Now()

			if _, err := ec.Encode(values1); err != nil {
				fmt.Printf("An error occured during encoding: %v\n", err)
				os.Exit(1)
			}

			ec.Dispose()

			after1 := time.Now()
			delta1 += after1.Sub(before1).Nanoseconds()

			if _, err := obs.Close(); err != nil {
				fmt.Printf("Error during close: %v\n", err)
				os.Exit(1)
			}

			ibs, _ := bitstream.NewDefaultInputBitStream(&bs, uint(size))
			ed := getDecoder(name, ibs)

			if ed == nil {
				os.Exit(1)
			}
			// Decode
			before2 := time.Now()

			if _, err := ed.Decode(values2); err != nil {
				fmt.Printf("An error occured during decoding: %v\n", err)
				os.Exit(1)
			}

			ed.Dispose()

			after2 := time.Now()
			delta2 += after2.Sub(before2).Nanoseconds()

			if _, err := ibs.Close(); err != nil {
				fmt.Printf("Error during close: %v\n", err)
				os.Exit(1)
			}

			// Sanity check
			for i := 0; i < size; i++ {
				if values1[i] != values2[i] {
					fmt.Printf("Error at index %v (%v<->%v)\n", i, values1[i], values2[i])
					break
				}
			}
		}

		bs.Close()

		fmt.Printf("Encode [ms]      : %d\n", delta1/1000000)
		fmt.Printf("Throughput [KB/s]: %d\n", (int64(iter*size))*1000000/delta1*1000/1024)
		fmt.Printf("Decode [ms]      : %d\n", delta2/1000000)
		fmt.Printf("Throughput [KB/s]: %d\n", (int64(iter*size))*1000000/delta2*1000/1024)
	}
}

File: java/src/test/java/kanzi/test/TestEntropyCodec.java
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.test;

import kanzi.BitStreamException;
import kanzi.entropy.BinaryEntropyDecoder;
import kanzi.entropy.BinaryEntropyEncoder;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.util.Random;
import kanzi.EntropyDecoder;
import kanzi.EntropyEncoder;
import kanzi.InputBitStream;
import kanzi.OutputBitStream;
import kanzi.bitstream.DebugOutputBitStream;
import kanzi.bitstream.DefaultInputBitStream;
import kanzi.bitstream.DefaultOutputBitStream;
import kanzi.entropy.ANSRangeDecoder;
import kanzi.entropy.ANSRangeEncoder;
import kanzi.entropy.CMPredictor;
import kanzi.entropy.ExpGolombDecoder;
import kanzi.entropy.ExpGolombEncoder;
import kanzi.entropy.FPAQPredictor;
import kanzi.entropy.HuffmanDecoder;
import kanzi.entropy.HuffmanEncoder;
import kanzi.entropy.PAQPredictor;
import kanzi.Predictor;
import kanzi.entropy.RangeDecoder;
import kanzi.entropy.RangeEncoder;
import kanzi.entropy.RiceGolombDecoder;
import kanzi.entropy.RiceGolombEncoder;
import kanzi.entropy.TPAQPredictor;
import org.junit.Assert;
import org.junit.Test;


public class TestEntropyCodec
{
    public static void main(String[] args)
    {
       if (args.length == 0)
       {
          args = new String[] { "-TYPE=ALL" };
       }

       String type = args[0].toUpperCase();

       if (type.startsWith("-TYPE=")) 
       {
           type = type.substring(6);
           System.out.println("Codec: " + type);

           if (type.equals("ALL"))
           {
              System.out.println("\n\nTestHuffmanCodec");
              
              if (testCorrectness("HUFFMAN") == false)
                System.exit(1);
             
              testSpeed("HUFFMAN");
              System.out.println("\n\nTestANS0Codec");
              
              if (testCorrectness("ANS0") == false)
                System.exit(1);
             
              testSpeed("ANS0");
              System.out.println("\n\nTestANS1Codec");
              
              if (testCorrectness("ANS1") == false)
                System.exit(1);
             
              testSpeed("ANS1");
              System.out.println("\n\nTestRangeCodec");
              
              if (testCorrectness("RANGE")== false)
                System.exit(1);
             
              testSpeed("RANGE");
              System.out.println("\n\nTestFPAQCodec");
              
              if (testCorrectness("FPAQ") == false)
                System.exit(1);
             
              testSpeed("FPAQ");
              System.out.println("\n\nTestCMCodec");
              
              if (testCorrectness("CM") == false)
                System.exit(1);
             
              testSpeed("CM");
              System.out.println("\n\nTestPAQCodec");
              
              if (testCorrectness("PAQ") == false)
                System.exit(1);
             
              testSpeed("PAQ");
              System.out.println("\n\nTestTPAQCodec");
              
              if (testCorrectness("TPAQ") == false)
                System.exit(1);
             
              testSpeed("TPAQ");
              System.out.println("\n\nTestExpGolombCodec");
              
              if (testCorrectness("EXPGOLOMB") == false)
                System.exit(1);
             
              testSpeed("EXPGOLOMB");
              System.out.println("\n\nTestRiceGolombCodec");
              
              if (testCorrectness("RICEGOLOMB") == false)
                System.exit(1);
             
              testSpeed("RICEGOLOMB");           
           }
           else
           {
              System.out.println("Test" + type + "Codec");
              testCorrectness(type);
              testSpeed(type);
           }
       }        
    }
    
   @Test
   public void testEntropy()
   {
      System.out.println("\n\nTestHuffmanCodec");
      Assert.assertTrue(testCorrectness("HUFFMAN"));
      //testSpeed("HUFFMAN");
      System.out.println("\n\nTestANS0Codec");
      Assert.assertTrue(testCorrectness("ANS0"));
      //testSpeed("ANS0");
      System.out.println("\n\nTestANS1Codec");
      Assert.assertTrue(testCorrectness("ANS1"));
      //testSpeed("ANS1");
      System.out.println("\n\nTestRangeCodec");
      Assert.assertTrue(testCorrectness("RANGE"));
      //testSpeed("RANGE");
      System.out.println("\n\nTestFPAQCodec");
      Assert.assertTrue(testCorrectness("FPAQ"));
      //testSpeed("FPAQ");
      System.out.println("\n\nTestCMCodec");
      Assert.assertTrue(testCorrectness("CM"));
      //testSpeed("CM");
      System.out.println("\n\nTestPAQCodec");
      Assert.assertTrue(testCorrectness("PAQ"));
      //testSpeed("PAQ");
      System.out.println("\n\nTestTPAQCodec");
      Assert.assertTrue(testCorrectness("TPAQ"));
      //testSpeed("TPAQ");
      System.out.println("\n\nTestExpGolombCodec");
      Assert.assertTrue(testCorrectness("EXPGOLOMB"));
      //testSpeed("EXPGOLOMB");
      System.out.println("\n\nTestRiceGolombCodec");
      Assert.assertTrue(testCorrectness("RICEGOLOMB"));
      //testSpeed("RICEGOLOMB");
   }
   
   
   private static Predictor getPredictor(String type)
   {
      if (type.equals("PAQ"))
         return new PAQPredictor();

      if (type.equals("TPAQ"))
         return new TPAQPredictor(null);

      if (type.equals("FPAQ"))
         return new FPAQPredictor();

      if (type.equals("CM"))
         return new CMPredictor();

      return null;
   }

    
   private static EntropyEncoder getEncoder(String name, OutputBitStream obs)
   {
      switch(name) 
      {
         case "FPAQ":
         case "CM":
         case "PAQ":
         case "TPAQ":
            return new BinaryEntropyEncoder(obs, getPredictor(name));

         case "HUFFMAN":
            return new HuffmanEncoder(obs);

         case "ANS0":
            return new ANSRangeEncoder(obs, 0);

         case "ANS1":
            return new ANSRangeEncoder(obs, 1);

         case "RANGE":
            return new RangeEncoder(obs);

         case "EXPGOLOMB":
            return new ExpGolombEncoder(obs, true);

         case "RICEGOLOMB":
            return new RiceGolombEncoder(obs, true, 4);

         default:
            System.out.println("No such entropy encoder: "+name);
            return null;
      }
   }


   private static EntropyDecoder getDecoder(String name, InputBitStream ibs)
   {
      switch(name) 
      {
         case "FPAQ":
         case "CM":
         case "PAQ":
         case "TPAQ":             
            Predictor pred = getPredictor(name);

            if (pred == null)
            {
               System.out.println("No such entropy decoder: "+name);
               return null;
            }

            return new BinaryEntropyDecoder(ibs, pred);

         case "HUFFMAN":
            return new HuffmanDecoder(ibs);

         case "ANS0":
            return new ANSRangeDecoder(ibs, 0);

         case "ANS1":
            return new ANSRangeDecoder(ibs, 1);

         case "RANGE":
            return new RangeDecoder(ibs);

         case "EXPGOLOMB":
            return new ExpGolombDecoder(ibs, true);

         case "RICEGOLOMB":
            return new RiceGolombDecoder(ibs, true, 4);

         default:
            System.out.println("No such entropy decoder: "+name);
            return null;
      }       
   }


   public static boolean testCorrectness(String name)
   {
      // Test behavior
      System.out.println("Correctness test for " + name);

      for (int ii=1; ii<20; ii++)
      {
          System.out.println("\n\nTest "+ii);

          try
          {
            byte[] values;
            Random random = new Random();

            if (ii == 3)
                 values = new byte[] { 0, 0, 32, 15, -4, 16, 0, 16, 0, 7, -1, -4, -32, 0, 31, -1 };
            else if (ii == 2)
                 values = new byte[] { 61, 77, 84, 71, 90, 54, 57, 38, 114, 111, 108, 101, 61, 112, 114, 101 };
            else if (ii == 4)
                 values = new byte[] { 65, 71, 74, 66, 76, 65, 69, 77, 74, 79, 68, 75, 73, 72, 77, 68, 78, 65, 79, 79, 78, 66, 77, 71, 64, 70, 74, 77, 64, 67, 71, 64 };
            else if (ii == 1)
            {
               values = new byte[32];

               for (int i=0; i<values.length; i++)
                    values[i] = (byte) 2; // all identical
            }
            else if (ii == 5)
            {
               values = new byte[32];

               for (int i=0; i<values.length; i++)
                    values[i] = (byte) (2 + (i & 1)); // 2 symbols
            }
            else
            {
               values = new byte[32];

               for (int i=0; i<values.length; i++)
                    values[i] = (byte) (64 + 3*ii + random.nextInt(ii+1));
            }

            System.out.println("Original:");

            for (int i=0; i<values.length; i++)
               System.out.print((values[i]&0xFF)+" ");

            System.out.println();
            System.out.println("\nEncoded:");
            ByteArrayOutputStream os = new ByteArrayOutputStream(16384);
            OutputBitStream obs = new DefaultOutputBitStream(os, 16384);
            DebugOutputBitStream dbgbs = new DebugOutputBitStream(obs, System.out);
            dbgbs.showByte(true);
            EntropyEncoder ec = getEncoder(name, dbgbs);

            if (ec == null)
               return false;

            ec.encode(values, 0, values.length);                
            ec.dispose();
            dbgbs.close();
            byte[] buf = os.toByteArray();
            InputBitStream ibs = new DefaultInputBitStream(new ByteArrayInputStream(buf), 1024);
            EntropyDecoder ed = getDecoder(name, ibs);

            if (ed == null)
               return false;

            System.out.println();
            System.out.println("\nDecoded:");
            boolean ok = true;
            byte[] values2 = new byte[values.length];
            ed.decode(values2, 0, values2.length);
            ed.dispose();
            ibs.close();

            try
            {
               for (int j=0; j<values2.length; j++)
               {
                  if (values[j] != values2[j])
                     ok = false;

                  System.out.print((values2[j]&0xFF)+" ");
               }
            }
            catch (BitStreamException e)
            {
               e.printStackTrace();
               return false;
            }

            System.out.println("\n"+((ok == true) ? "Identical" : "Different"));
         }
         catch (Exception e)
         {
            e.printStackTrace();
            return false;
         }        
      }
          
      return true;      
   }

    
   public static void testSpeed(String name)
   {
      // Test speed
      System.out.println("\n\nSpeed test for " + name);
      int[] repeats = { 3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5, 8, 9, 7, 9, 3 };
      final int size = 500000;
      final int iter = 100;
      Random random = new Random();

      for (int jj=0; jj<3; jj++)
      {
          System.out.println("\nTest "+(jj+1));
          byte[] values1 = new byte[size];
          byte[] values2 = new byte[size];
          long delta1 = 0, delta2 = 0;

          for (int ii=0; ii<iter; ii++)
          {
              int idx = 0;

              for (int i=0; i<size; i++)
              {
                  int i0 = i;
                  int len = repeats[idx];
                  idx = (idx + 1) & 0x0F;
                  byte b = (byte) random.nextInt(256);

                  if (i0+len >= size)
                      len = size-i0-1;

                  for (int j=i0; j<i0+len; j++)
                  {
                     values1[j] = b;
                     i++;
                  }
              }

              // Encode
              ByteArrayOutputStream os = new ByteArrayOutputStream(size*2);
              OutputBitStream obs = new DefaultOutputBitStream(os, size);
              EntropyEncoder ec = getEncoder(name, obs);

              if (ec == null)
                 System.exit(1);

              long before1 = System.nanoTime();

              if (ec.encode(values1, 0, values1.length) < 0)
              {
                 System.out.println("Encoding error");
                 System.exit(1);
              }

              ec.dispose();
              long after1 = System.nanoTime();
              delta1 += (after1 - before1);
              obs.close();

              // Decode
              byte[] buf = os.toByteArray();
              InputBitStream ibs = new DefaultInputBitStream(new ByteArrayInputStream(buf), size);
              EntropyDecoder ed = getDecoder(name, ibs);

              if (ed == null)
                 System.exit(1);

              long before2 = System.nanoTime();

              if (ed.decode(values2, 0, size) < 0)
              {
                 System.out.println("Decoding error");
                 System.exit(1);
              }

              ed.dispose();
              long after2 = System.nanoTime();
              delta2 += (after2 - before2);
              ibs.close();

              // Sanity check
              for (int i=0; i<size; i++)
              {
                 if (values1[i] != values2[i])
                 {
                    System.out.println("Error at index "+i+" ("+values1[i]
                            +"<->"+values2[i]+")");
                    break;
                 }
              }
          }

          final long prod = (long) iter * (long) size;
          System.out.println("Encode [ms]       : " + delta1/1000000L);
          System.out.println("Throughput [KB/s] : " + prod * 1000000L / delta1 * 1000L / 1024L);
          System.out.println("Decode [ms]       : " + delta2/1000000L);
          System.out.println("Throughput [KB/s] : " + prod * 1000000L / delta2 * 1000L / 1024L);
      }
  }
}
